core/ExecutePipe.cpp

```diff
@@ -180,43 +206,108 @@ namespace olympia
         uint32_t credits_to_send = 0;
-        while(it != ready_queue_.end()) {
-            if((*it)->getUniqueID() >= uint64_t(criteria)) {
-                ready_queue_.erase(it++);
+
+        auto issue_queue_iter = issue_queue_.begin();
+        while (issue_queue_iter != issue_queue_.end())
+        {
+            auto delete_iter = issue_queue_iter++;
+
+            auto inst_ptr = *delete_iter;
+
+            // Remove flushed instruction from issue queue and clear scoreboard callbacks
+            if (criteria.includedInFlush(inst_ptr))
+            {
+                issue_queue_.erase(delete_iter);
+
+                scoreboard_views_[reg_file_]->clearCallbacks(inst_ptr->getUniqueID());
+
                 ++credits_to_send;
-            }
-            else {
-                ++it;
+
+                ILOG("Flush Instruction ID: " << inst_ptr->getUniqueID() << " from issue queue");
             }
         }
+
         if(credits_to_send) {
             out_scheduler_credits_.send(credits_to_send, 0);
+
+            ILOG("Flush " << credits_to_send << " instructions in issue queue!");
+        }
+
+        // Flush instructions in the ready queue
+        auto ready_queue_iter = ready_queue_.begin();
+        while (ready_queue_iter != ready_queue_.end())
+        {
+            auto delete_iter = ready_queue_iter++;
+            auto inst_ptr = *delete_iter;
+            if (criteria.includedInFlush(inst_ptr))
+            {
+                ready_queue_.erase(delete_iter);
+                ILOG("Flush Instruction ID: " << inst_ptr->getUniqueID() << " from ready queue");
+            }
         }
 
         // Cancel outstanding instructions awaiting completion and
         // instructions on their way to issue
-        auto cancel_critera = [criteria](const InstPtr & inst) -> bool {
-            if(inst->getUniqueID() >= uint64_t(criteria)) {
-                return true;
-            }
-            return false;
+        auto flush = [criteria](const InstPtr & inst) -> bool {
+            return criteria.includedInFlush(inst);
         };
-        complete_inst_.cancelIf(cancel_critera);
         issue_inst_.cancel();
-
-        if(complete_inst_.getNumOutstandingEvents() == 0) {
+        complete_inst_.cancelIf(flush);
+        execute_inst_.cancelIf(flush);
+        if(execute_inst_.getNumOutstandingEvents() == 0) {
             unit_busy_ = false;
             collected_inst_.closeRecord();
+            if (!ready_queue_.empty()) {
+                // issue non-flushed instructions
+                issue_inst_.schedule(sparta::Clock::Cycle(1));
+            }
+        }
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////
+    // Regular Function/Subroutine Call
+    ////////////////////////////////////////////////////////////////////////////////
+
+    // Append instruction into issue queue
+    void ExecutePipe::appendIssueQueue_(const InstPtr & inst_ptr)
```

Comment:
The [Sparta Buffer class](https://github.com/sparcians/map/blob/8b16b216de9960cacdc22d78820165ab86d6de58/sparta/sparta/resources/Buffer.hpp) might be good to use for the issue queue. It allows for append and erase from anywhere in the buffer. It has its own implementation of the two methods you have here (`appendIssueQueue_` and `popIssueQueue_`) so you wouldn't need to provide your own implementation.

Reply:
Might be something @AaronGChan can pick up in the issue queue changes